sync/sync_tail.go

```diff
@@ -0,0 +1,165 @@
+		if tailHeight <= s.store.Height() {
+			tail, err = s.store.GetByHeight(ctx, tailHeight)
+		}
+		if err != nil || tailHeight != tail.Height() {
+			log.Debugw("tail height not available locally, fetching...", "height", tailHeight)
+			tail, err = s.getter.GetByHeight(ctx, tailHeight)
+			if err != nil {
+				return tail, fmt.Errorf("getting SyncFromHeight tail(%d): %w", tailHeight, err)
+			}
+			fetched = true
+		}
+	} else if tailHash == nil && tailHeight == 0 {
+		if tail.IsZero() {
+			// no previously known Tail available - estimate solely on Head
+			estimatedHeight := estimateTail(head, s.Params.blockTime, s.Params.TrustingPeriod)
+			tail, err = s.getter.GetByHeight(ctx, estimatedHeight)
+			if err != nil {
+				return tail, fmt.Errorf("getting estimated tail(%d): %w", tailHeight, err)
+			}
+			fetched = true
+		} else {
+			// have a known Tail - estimate basing on it.
+			cutoffTime := head.Time().UTC().Add(-s.Params.PruningWindow)
+			diff := cutoffTime.Sub(tail.Time().UTC())
+			if diff <= 0 {
+				// current tail is relevant as is
+				return tail, nil
+			}
+			log.Debugw("current tail is beyond pruning window", "current_height", tail.Height(), "diff", diff.String())
+
+			toDeleteEstimate := uint64(diff / s.Params.blockTime) //nolint:gosec
+			estimatedNewTail := tail.Height() + toDeleteEstimate
+
+			for {
+				tail, err = s.store.GetByHeight(ctx, estimatedNewTail)
+				if err != nil {
+					log.Errorw("getting estimated tail from store", "height", estimatedNewTail, "error", err)
+					return tail, err
+				}
+				if tail.Time().UTC().Compare(cutoffTime) <= 0 {
+					// tail before or equal to cutoffTime
+					break
+				}
+
+				estimatedNewTail++
+			}
+
+			log.Debugw("estimated new tail", "new_height", tail.Height())
+		}
+	}
+
+	if fetched {
+		if err := s.store.Append(ctx, tail); err != nil {
+			return tail, fmt.Errorf("appending tail header: %w", err)
+		}
+	}
+
+	if err := s.moveTail(ctx, tail); err != nil {
+		return tail, fmt.Errorf("moving tail: %w", err)
+	}
+
+	return tail, nil
+}
+
+// moveTail moves the Tail to be the given header.
+// It will prune the store if the new Tail is higher than the old one or
+// sync up if the new Tail is lower than the old one.
+func (s *Syncer[H]) moveTail(ctx context.Context, to H) error {
+	from, err := s.store.Tail(ctx)
+	if errors.Is(err, header.ErrEmptyStore) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case from.Height() < to.Height():
+		log.Infof("move tail up from %d to %d, pruning the diff...", from.Height(), to.Height())
+		err := s.store.DeleteTo(ctx, to.Height())
+		if err != nil {
+			return fmt.Errorf(
+				"deleting headers up to newly configured tail(%d): %w",
+				to.Height(),
+				err,
+			)
+		}
+	case from.Height() > to.Height():
+		log.Infof("move tail down from %d to %d, syncing the diff...", from.Height(), to.Height())
+
+		// TODO(@Wondertan): This works but it assumes this code is only run before syncing routine starts.
```

Comment:
Yeah afaict we would never "grow" tail in runtime -- this would only happen on a restart for a user who wants to specify "lower" tail height 

and this would be hit before a new subjHead would be set which is what would trigger a sync job in the forward direction.

Reply:
not actionable